### ISSUE TITLE
[webview_flutter] Gestures stop working

### ISSUE THREAD

**Issue by ponytony** [flutter/flutter#56860]
## Steps to Reproduce

1. clone 'https://github.com/ponytony/flutter-webview-demo', switch to demo branch.
2. run the project
3. open an flutter application on ios simulator(ios version is 13.4.1), the flutter page have an webview component which url is 'localhost:8080'
4. touch the 'test' button
5. select a time
6. click cancel
7.  Repeat the above actions about ten times

**Actual results:** No click or touch event be triggered,but dom's click method can be call.

**why I submit the issue:** 
1. the bug is occur on ios 13.4.1
2. another component also occur the bug
3. no console log;
4. run the page on safari is correct

**flutter doctor**

```
[✓] Flutter (Channel unknown, v1.12.13+hotfix.5, on Mac OS X 10.15.4 19E287, locale zh-Hans-CN)
    • Flutter version 1.12.13+hotfix.5 at /Users/wengu/Desktop/flutter
    • Framework revision 27321ebbad (5 months ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/wengu/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/wengu/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] Connected device (1 available)
    • iPhone SE (2nd generation) • 0BC83A31-1F55-4E79-8E9F-860C78FCD113 • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-4 (simulator)

• No issues found!
```

**flutter analyze:**
```
no issue found
```

**Comment by VladyslavBondarenko** [comment]
Hi @ponytony 
I can't see any logs on clicks in the console where I run `npm run serve`. 
Does your issue persist with current dev channel `1.19.0-0.0.pre` or current stable `1.17.0`, what `webview_flutter` are you using?

**Comment by ponytony** [comment]
> Hi @ponytony
> I can't see any logs in the console where I run `npm run serve`, when I click in browser as well.
> Does your issue persist with current dev channel `1.19.0-0.0.pre` or current stable `1.17.0`, what `webview_flutter` are you using?

I also dont see any log in the console.the channel is stable chennel v1.12.13+hotfix.5, and the webview flutter is official plugin.

**Comment by ponytony** [comment]
> Hi @ponytony
> I can't see any logs in the console where I run `npm run serve`, when I click in browser as well.
> Does your issue persist with current dev channel `1.19.0-0.0.pre` or current stable `1.17.0`, what `webview_flutter` are you using?

you can see the 'Steps to Reproduce', step 4,5,6 is important

**Comment by VladyslavBondarenko** [comment]
@ponytony 
> webview flutter is official plugin.

Oops, I missed the word `version` in my question.

> No click or touch event be triggered

what events are you expecting to have triggered?

> run the page on safari is correct

what is the difference with running in flutter app in webview?

**Comment by ponytony** [comment]
> @ponytony
> 
> > webview flutter is official plugin.
> 
> Oops, I missed the word `version` in my question.
> 
> > No click or touch event be triggered
> 
> what events are you expecting to have triggered?
> 
> > run the page on safari is correct
> 
> what is the difference with running in flutter app in webview?

version of webview_flutter is 0.3.13.ummm, i want to trigger any event, but the page looks like be locked when the bug happen, i cannot do anything；

**Comment by VladyslavBondarenko** [comment]
@ponytony 
try with the last version of webview_flutter 0.3.21.
It's still not clear to me what is the difference between running in browser and in flutter app with webview

**Comment by ponytony** [comment]
> @ponytony
> try with the last version of webview_flutter 0.3.21.
> It's still not clear to me what is the difference between running in browser and in flutter app with webview

I've upload video to github, repo is https://github.com/ponytony/flutter-webview-demo', branch is demo

**Comment by VladyslavBondarenko** [comment]
The freeze reproduces with webview on iOS (not with Android), but doesn't reproduce when opening the site in mobile browser. Tried the last version `webview_flutter: ^0.3.22+1`.

@ponytony For future, to avoid downloading your repository and running it locally by other people to reproduce would be better if you deploy your site and provide a code sample containing WebView with the link on the site

<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel dev, 1.19.0-1.0.pre, on Mac OS X 10.15.4 19E287, locale en-GB)
    • Flutter version 1.19.0-1.0.pre at /Users/nevercode/dev/flutter_dev
    • Framework revision 456d80b9dd (11 days ago), 2020-05-11 11:45:03 -0400
    • Engine revision d96f962ca2
    • Dart version 2.9.0 (build 2.9.0-7.0.dev 092ed38a87)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.45.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.2

[✓] Connected device (6 available)
    • Redmi Note 7       • 6345c469                                 • android-arm64  • Android 9 (API 28)
    • Nevercode’s iPhone • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.5
    • iPhone 11 Pro Max  • 9F9AFD9B-7F22-468E-8096-B47E993E0619     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)
    • macOS              • macOS                                    • darwin-x64     • Mac OS X 10.15.4 19E287
    • Web Server         • web-server                               • web-javascript • Flutter Tools
    • Chrome             • chrome                                   • web-javascript • Google Chrome 83.0.4103.61

• No issues found!
```

</details>

Related to https://github.com/flutter/flutter/issues/53490. Follow up on it and check your issue once the mentioned is fixed

**Comment by Voxar** [comment]
Hi! I might have some details here. It's related to video i my case but I suspect the cause might be the same. 

_TL;DR: I think the mechanism flutter uses to track touches in platform views does not receive all touch events causing its state to lose sync with reality and ends up blocking further touches._ 


We have modified webiew_flutter to add `WKWebViewConfiguration.allowsInlineMediaPlayback = true` for inline video support and get this issue frequently. 

I created a sample flutter project that just opens a youtube video in the modified WebView and can there reproduce this issue very easily by, after starting the video, simply tapping the time slider thumb once, then tapping anywhere else on the inlined video. After this no further taps are registered for the WebView.


I have so far seen the following effects which I suspect may be related:

Creating a normal xcode ios sample project and attaching a gesture recogniser to the WKWebView one can see that tapping the video time slider thumb results in a call to `touchesBegan` but not `touchesEnded`.

After reaching the faulty state while debugging the flutter project in xcode, one can find the reference to the `FlutterTouchInterceptingView` that flutter uses to intercept touches via gesture recognisers; `DelayingGestureRecognizer` and `ForwardingGestureRecognizer`. Inspecting the `ForwardingGestureRecognizer._currentTouchPointersCount` value reveals that it is stuck at 1 while it should be 0, and thus the recogniser is in an unexpected state.

Calling `[ForwardingGestureRecognizer touchesCancelled:nil withEvent:nil]` from the debugger restores the `ForwardingGestureRecognizer` to the proper state and the webview receives touches again.


For details on how flutter handles (and blocks) touches see
https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L892



I haven't been able to run @ponytony's example but if you would like to confirm the state of the gesture recogniser you can stop the debugger and run this little line in gdb (mine was stuck in objc mode at the time. Probably only works reliably in a simple project)

```obj-c
po [[[[UIApplication.sharedApplication.keyWindow.rootViewController.view.subviews valueForKeyPath:@"@unionOfArrays.gestureRecognizers"] filteredArrayUsingPredicate: [NSPredicate predicateWithBlock: ^BOOL(id item, id b){ return (BOOL)[item isKindOfClass: [ForwardingGestureRecognizer class]];}]] firstObject] valueForKey:@"_currentTouchPointersCount"]
```

**Comment by Voxar** [comment]
Here's a hack workaround that works for us. Hope this helps

https://gist.github.com/Voxar/5dbef0b76cc8cb55d2fe74a492d41c54

**Comment by TahaTesser** [comment]
Hi @ponytony  @Voxar 
The project provided no longer works, can you a new reproducible code or simpler way of reproducing this which doesn't require nodejs.

Thank you

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
